File: modules/identity-owner/android/src/main/java/org/nostrocket/hypergolic/identityowner/IdentityOwnerModule.kt
package org.nostrocket.hypergolic.identityowner

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class IdentityOwnerModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("HypergolicIdentityOwner")
    Function("claimIdentityOwner") {
      val onMainJavaScriptThread = runCatching { appContext.runtime.reactContext?.isOnJSQueueThread == true }.getOrDefault(false)
      IdentityOwnerClaim.contexts.claim(appContext, onMainJavaScriptThread)
    }
    OnDestroy { IdentityOwnerClaim.contexts.retireIfOwned(appContext) }
  }
}

File: modules/identity-owner/ios/IdentityOwnerModule.swift
import ExpoModulesCore

public final class IdentityOwnerModule: Module {
  public func definition() -> ModuleDefinition {
    Name("HypergolicIdentityOwner")
    Function("claimIdentityOwner") { () -> Bool in
      guard IdentityOwnerClaim.claim() else { NativeIdentityOwnerContext.retire(); return false }
      // The actual main AppContext/runtime comes from native module construction, never JS input.
      guard let appContext, let runtime = try? appContext.runtime, runtime.isOnJavaScriptThread(),
        NativeIdentityOwnerContext.admit(appContext) else { NativeIdentityOwnerContext.retire(); return false }
      return true
    }
  }
}
